## Time Series in Pandas

Exploring about DateTime function applied in PANDAS which is used for analyse Time series data

In [2]:
#Import packages and datasets

import pandas as pd
import numpy as np

In [3]:
air_df = pd.read_csv('https://raw.githubusercontent.com/insaid2018/Term-1/master/Data/Casestudy/AirQualityUCI.csv') 
air_df.head()

Date      Time  CO(GT)  PT08.S1(CO)  NMHC(GT)  C6H6(GT)  \
0  3/10/2004  18:00:00     2.6         1360       150      11.9   
1  3/10/2004  19:00:00     2.0         1292       112       9.4   
2  3/10/2004  20:00:00     2.2         1402        88       9.0   
3  3/10/2004  21:00:00     2.2         1376        80       9.2   
4  3/10/2004  22:00:00     1.6         1272        51       6.5   

   PT08.S2(NMHC)  NOx(GT)  PT08.S3(NOx)  NO2(GT)  PT08.S4(NO2)  PT08.S5(O3)  \
0           1046      166          1056      113          1692         1268   
1            955      103          1174       92          1559          972   
2            939      131          1140      114          1555         1074   
3            948      172          1092      122          1584         1203   
4            836      131          1205      116          1490         1110   

      T    RH      AH  
0  13.6  48.9  0.7578  
1  13.3  47.7  0.7255  
2  11.9  54.0  0.7502  
3  11.0  60.0  0.7867  
4  11.2  59.6  0.7888

In [4]:
air_df.shape

(9357, 15)

In [5]:
air_df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 9357 entries, 0 to 9356
Data columns (total 15 columns):
Date             9357 non-null object
Time             9357 non-null object
CO(GT)           9357 non-null float64
PT08.S1(CO)      9357 non-null int64
NMHC(GT)         9357 non-null int64
C6H6(GT)         9357 non-null float64
PT08.S2(NMHC)    9357 non-null int64
NOx(GT)          9357 non-null int64
PT08.S3(NOx)     9357 non-null int64
NO2(GT)          9357 non-null int64
PT08.S4(NO2)     9357 non-null int64
PT08.S5(O3)      9357 non-null int64
T                9357 non-null float64
RH               9357 non-null float64
AH               9357 non-null float64
dtypes: float64(5), int64(8), object(2)
memory usage: 1.1+ MB


In [6]:
air_df.describe()

CO(GT)  PT08.S1(CO)     NMHC(GT)     C6H6(GT)  PT08.S2(NMHC)  \
count  9357.000000  9357.000000  9357.000000  9357.000000    9357.000000   
mean    -34.207524  1048.990061  -159.090093     1.865683     894.595276   
std      77.657170   329.832710   139.789093    41.380206     342.333252   
min    -200.000000  -200.000000  -200.000000  -200.000000    -200.000000   
25%       0.600000   921.000000  -200.000000     4.000000     711.000000   
50%       1.500000  1053.000000  -200.000000     7.900000     895.000000   
75%       2.600000  1221.000000  -200.000000    13.600000    1105.000000   
max      11.900000  2040.000000  1189.000000    63.700000    2214.000000   

           NOx(GT)  PT08.S3(NOx)      NO2(GT)  PT08.S4(NO2)  PT08.S5(O3)  \
count  9357.000000   9357.000000  9357.000000   9357.000000  9357.000000   
mean    168.616971    794.990168    58.148873   1391.479641   975.072032   
std     257.433866    321.993552   126.940455    467.210125   456.938184   
min    -200.000000   -200.000000  -200.000000   -200.000000  -200.000000   
25%      50.000000    637.000000    53.000000   1185.000000   700.000000   
50%     141.000000    794.000000    96.000000   1446.000000   942.000000   
75%     284.000000    960.000000   133.000000   1662.000000  1255.000000   
max    1479.000000   2683.000000   340.000000   2775.000000  2523.000000   

                 T           RH           AH  
count  9357.000000  9357.000000  9357.000000  
mean      9.778305    39.485380    -6.837604  
std      43.203623    51.216145    38.976670  
min    -200.000000  -200.000000  -200.000000  
25%      10.900000    34.100000     0.692300  
50%      17.200000    48.600000     0.976800  
75%      24.100000    61.900000     1.296200  
max      44.600000    88.700000     2.231000

#### As there's -200 in min and -200 in NMHC(GT) as well which is quite weird in data which is having sensor based readings but if 

#### sensor is not capturing readings properly only then -200 is being captured automatically instead of capture "NAN" or missing cell. 

In [18]:
air_df[air_df['NMHC(GT)'] == -200]['NMHC(GT)'].count()

8443

In [23]:
air_df.drop(['NMHC(GT)'], axis =1, inplace= True)

#### Now we will do changes in time and date formats

In [14]:
air_df['Date'] = pd.to_datetime(air_df['Date'], format='%m/%d/%y')

In [15]:
air_df['Time'] = pd.to_datetime(air_df['Time'], format='%H:%M:%S')

In [24]:
air_df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 9357 entries, 0 to 9356
Data columns (total 14 columns):
Date             9357 non-null datetime64[ns]
Time             9357 non-null datetime64[ns]
CO(GT)           9357 non-null float64
PT08.S1(CO)      9357 non-null int64
C6H6(GT)         9357 non-null float64
PT08.S2(NMHC)    9357 non-null int64
NOx(GT)          9357 non-null int64
PT08.S3(NOx)     9357 non-null int64
NO2(GT)          9357 non-null int64
PT08.S4(NO2)     9357 non-null int64
PT08.S5(O3)      9357 non-null int64
T                9357 non-null float64
RH               9357 non-null float64
AH               9357 non-null float64
dtypes: datetime64[ns](2), float64(5), int64(7)
memory usage: 1023.5 KB


In [25]:
air_df.head()

Date                Time  CO(GT)  PT08.S1(CO)  C6H6(GT)  \
0 2004-03-10 1900-01-01 18:00:00     2.6         1360      11.9   
1 2004-03-10 1900-01-01 19:00:00     2.0         1292       9.4   
2 2004-03-10 1900-01-01 20:00:00     2.2         1402       9.0   
3 2004-03-10 1900-01-01 21:00:00     2.2         1376       9.2   
4 2004-03-10 1900-01-01 22:00:00     1.6         1272       6.5   

   PT08.S2(NMHC)  NOx(GT)  PT08.S3(NOx)  NO2(GT)  PT08.S4(NO2)  PT08.S5(O3)  \
0           1046      166          1056      113          1692         1268   
1            955      103          1174       92          1559          972   
2            939      131          1140      114          1555         1074   
3            948      172          1092      122          1584         1203   
4            836      131          1205      116          1490         1110   

      T    RH      AH  
0  13.6  48.9  0.7578  
1  13.3  47.7  0.7255  
2  11.9  54.0  0.7502  
3  11.0  60.0  0.7867  
4  11.2  59.6  0.7888

In [27]:
# Extracting year Column from Date

Year = air_df.Date.dt.year
Year

0       2004
1       2004
2       2004
3       2004
4       2004
        ... 
9352    2005
9353    2005
9354    2005
9355    2005
9356    2005
Name: Date, Length: 9357, dtype: int64

In [28]:
# Extracting month Column from Date

Month = air_df.Date.dt.month
Month

0       3
1       3
2       3
3       3
4       3
       ..
9352    4
9353    4
9354    4
9355    4
9356    4
Name: Date, Length: 9357, dtype: int64

In [30]:
Month.unique()

array([ 3,  4,  5,  6,  7,  8,  9, 10, 11, 12,  1,  2], dtype=int64)

In [33]:
Month.nunique()

12

In [31]:
# Extracting Day Column from Date

Day = air_df.Date.dt.day
Day

0       10
1       10
2       10
3       10
4       10
        ..
9352     4
9353     4
9354     4
9355     4
9356     4
Name: Date, Length: 9357, dtype: int64

In [32]:
Day.nunique()

31

In [34]:
# Extracting Day Column from Date

Week = air_df.Date.dt.week
Week

0       11
1       11
2       11
3       11
4       11
        ..
9352    14
9353    14
9354    14
9355    14
9356    14
Name: Date, Length: 9357, dtype: int64

In [35]:
Week.unique()

array([11, 12, 13, 14, 15, 16, 17, 18, 19, 20, 21, 22, 23, 24, 25, 26, 27,
       28, 29, 30, 31, 32, 33, 34, 35, 36, 37, 38, 39, 40, 41, 42, 43, 44,
       45, 46, 47, 48, 49, 50, 51, 52, 53,  1,  2,  3,  4,  5,  6,  7,  8,
        9, 10], dtype=int64)

In [36]:
Week.nunique()

53

The index of week days are provided below.
0 = Monday
1 = Tuesday
2 = Wednesday
3 = Thursday
4 = Friday
5 = Saturday
6 = Sunday

In [38]:
Day_of_week =air_df.Date.dt.dayofweek
Day_of_week

0       2
1       2
2       2
3       2
4       2
       ..
9352    0
9353    0
9354    0
9355    0
9356    0
Name: Date, Length: 9357, dtype: int64

In [39]:
Weekday_name = air_df.Date.dt.weekday_name
Weekday_name

0       Wednesday
1       Wednesday
2       Wednesday
3       Wednesday
4       Wednesday
          ...    
9352       Monday
9353       Monday
9354       Monday
9355       Monday
9356       Monday
Name: Date, Length: 9357, dtype: object

In [40]:
Day_of_year =air_df.Date.dt.dayofyear
Day_of_year

0       70
1       70
2       70
3       70
4       70
        ..
9352    94
9353    94
9354    94
9355    94
9356    94
Name: Date, Length: 9357, dtype: int64

In [42]:
Hour = air_df.Time.dt.hour
Hour

0       18
1       19
2       20
3       21
4       22
        ..
9352    10
9353    11
9354    12
9355    13
9356    14
Name: Time, Length: 9357, dtype: int64

In [43]:
Minute = air_df.Time.dt.minute
Minute

0       0
1       0
2       0
3       0
4       0
       ..
9352    0
9353    0
9354    0
9355    0
9356    0
Name: Time, Length: 9357, dtype: int64

In [49]:
time_stamp = pd.to_datetime("18:00:00", format='%H:%M:%S')

air_df[air_df['Time'] < time_stamp].shape

(7017, 14)

In [51]:
time_stamp = pd.to_datetime("09:00:00", format='%H:%M:%S')

air_df[air_df['Time'] > time_stamp].shape

(5457, 14)

There are 7017 records where time is before 1800 hrs.

Performing Conditional operations on Date. 
Lets measure the number of records before 01/01/2005

In [58]:
date_stamp = pd.to_datetime("01/01/2005", format='%d/%m/%Y')
air_df[air_df['Date'] < date_stamp].shape

(7110, 14)

We have seen that by using datetime format we can extract information like year, month, hour, minute etc from time.

We can even retrieve name of the day and day of the year type information.

All these things help to better understand our data which in turn help in drawing important conclusions and insights.